serving/samples: update helloworld-go to go:1.13 and go modules

grayside opened the pull request:
First step toward #1686 

## Proposed Changes

- Upgrades from go 1.12 to go 1.13
- Upgrades to use go modules
- Pins alpine image to 3.10
- Tweaks working directory to /app
- Rename binary in container from `helloworld` to `server`
- Use `go mod download` to facilitate a layer cache of go deps.
- Use `-mod=readonly` flag on builds for strictness of container builds
- Some minor style changes around syntax and comments

grayside commented:
/assign @ahmetb

grayside commented:
/test pull-knative-docs-unit-tests

ahmetb commented on docs/serving/samples/hello-world/helloworld-go/Dockerfile:
```diff
@@ -1,24 +1,25 @@
 # This is based on Debian and sets the GOPATH to /go.
 # https://hub.docker.com/_/golang
-FROM golang:1.12 as builder
+FROM golang:1.13 as builder
```
Any particular reason not to use golang:1?

We have to keep updating this every 6 months otherwise.

grayside replied:
Given that sometimes the minor version has significant change to best practices (e.g., modules) I think the clarity is worthwhile for the reader and motivation to update.

ahmetb commented on docs/serving/samples/hello-world/helloworld-go/Dockerfile:
```diff
@@ -1,24 +1,25 @@
+# https://hub.docker.com/_/alpine
 # https://docs.docker.com/develop/develop-images/multistage-build/#use-multi-stage-builds
-FROM alpine
+FROM alpine:3.10
```
Ditto. https://twitter.com/ahmetb/status/1164211673210609665?s=21

grayside replied:
```suggestion
FROM alpine:3
```

ahmetb commented on docs/serving/samples/hello-world/helloworld-go/Dockerfile:
```diff
@@ -1,24 +1,25 @@
 # Build the command inside the container.
 # (You may fetch or manage dependencies here,
 # either manually or with a tool like "godep".)
 RUN CGO_ENABLED=0 GOOS=linux go build -v -o helloworld
```
Imo “-o hello-server” is more descriptive. Or “-o app”

In such tutorials we end up with naming everything hello, so it’s often unclear what they refer to.

grayside replied:
Agreed.

tbpg commented on docs/serving/samples/hello-world/helloworld-go/README.md:
```diff
@@ -72,27 +72,28 @@ cd knative-docs/docs/serving/samples/hello-world/helloworld-go
-   FROM golang:1.12 as builder
-
+   FROM golang:1.13 as builder
+   
```
Extra whitespace (here and in a few other places).

grayside replied:
The Dockerfile spacing and comments are designed to emphasize a pipeline of steps and provide context on how the container is assembled.

grayside replied:
Ah, README whitespace. Addressing.

tbpg commented on docs/serving/samples/hello-world/helloworld-go/README.md:
```diff
@@ -72,27 +72,28 @@ cd knative-docs/docs/serving/samples/hello-world/helloworld-go
+   COPY . ./
+   
    # Build the command inside the container.
    # (You may fetch or manage dependencies here,
```
Delete this comment?

tbpg commented on docs/serving/samples/hello-world/helloworld-go/Dockerfile:
```diff
@@ -1,24 +1,25 @@
+COPY . ./
 
 # Build the command inside the container.
 # (You may fetch or manage dependencies here,
```
Delete this comment?

grayside replied:
Simplified the comment. The Dockerfile spacing and comments are designed to emphasize a pipeline of steps and provide context on how the container is assembled.

tbpg commented on docs/serving/samples/hello-world/helloworld-go/Dockerfile:
```diff
@@ -1,24 +1,25 @@
-WORKDIR /go/src/github.com/knative/docs/helloworld
-COPY . .
+WORKDIR /app
+COPY . ./
```
I wonder if this should do the `go mod download` trick, since many people will copy this? I know it's been decided on in the past, but consider adding it.

(I'm not attached to adding it - definitely makes things more complex, but it might be worth it.)

grayside replied:
Added go mod download and modified go build to use -mod=readonly for strictness of the build.

grayside commented:
/test pull-knative-docs-integration-tests

chaodaiG commented:
let me TAL

Fredy-Z commented:
> Looking at the error, I see:
> 
> ```
>             Step 3/10 : COPY go.* ./
>             COPY failed: no source files were specified
> ```
> 
> However, there is plainly a go.mod file in this sample and it builds locally just fine. Is it possible the test rig itself is skipping a file or missing the addition of go.mod somehow?
> 
> @chaodaiG could you take a look?

@grayside you'll need to add `go.mod` in https://github.com/knative/docs/blob/533e6b7bf310bcce3ccfb628e0e736f826235844/test/sampleapp/config.yaml#L14.
Also as I tried in https://github.com/knative/docs/pull/1787, you'll also need to remove `-mod=readonly` from the `go build` command.

chaodaiG commented:
Thank you @Fredy-Z for looking into this!

Fredy-Z commented on docs/serving/samples/hello-world/helloworld-go/go.mod:
```diff
@@ -0,0 +1,3 @@
+module github.com/knative/docs/docs/serving/samples/hello-world/helloworld-go
+
+go 1.13
```
Missed an empty line after this line. Otherwise `go build` will not work with `-mod=readonly`.

Fredy-Z commented:
> Thanks @Fredy-Z,
> 
> I want to make sure I understand what's happening here.
> 
> 1. Why can't I use -mod=readonly? That seems like it should be the best practice in a container, especially where the deps are downloaded via `go mod download`? In this case, the test framework appears to be constraining the implementation of the app in unexpected ways.
> 2. Why doesn't the test process copy all the files, perhaps excluding just .dockerignore? In this case, the test framework seems to be defining unexpected conventions.

@grayside sorry for the late reply.

For 1, I double checked your PR, this reason that you cannot have `-mod=readonly` is because `go.mod` file needs to be ended with an empty line, see https://github.com/knative/docs/pull/1774#discussion_r324506203.
For 2, I'm not sure what's the historic reason for only copying specific files, to me it's totally fine to just copy all files. @chaodaiG can you take a look?

chaodaiG commented:
> > Thanks @Fredy-Z,
> > I want to make sure I understand what's happening here.
> > 
> > 1. Why can't I use -mod=readonly? That seems like it should be the best practice in a container, especially where the deps are downloaded via `go mod download`? In this case, the test framework appears to be constraining the implementation of the app in unexpected ways.
> > 2. Why doesn't the test process copy all the files, perhaps excluding just .dockerignore? In this case, the test framework seems to be defining unexpected conventions.
> 
> @grayside sorry for the late reply.
> 
> For 1, I double checked your PR, this reason that you cannot have `-mod=readonly` is because the `go.mod` file is not ended with an empty line, see [#1774 (comment)](https://github.com/knative/docs/pull/1774#discussion_r324506203).
> For 2, I'm not sure what's the historic reason for only copying specific files, to me it's totally fine to just copy all files. @chaodaiG can you take a look?

A piece of background, I have observed that all files needed for tests are already inlined in `README.md` files, so there is a unit test verifying the consistency between committed code and README, and the unit test goes through all files needs to be copied and verify if they are identical with `README.md`. See test here: https://github.com/knative/docs/blob/533e6b7bf310bcce3ccfb628e0e736f826235844/test/sampleconsistency/sampleconsistency_test.go#L96

grayside commented:
For `-mod=readonly`, the container builds fine for me locally as-is, not sure what I'd change there. Why is go build in the container different in CI here?

For file-copy: It makes sense to me that we test that files we include in the README are consistent with source code, but that should not determine what is copied. For example, there are lockfiles such as `package-lock.json` for Node.js which need to be included in the build.

grayside commented:
Updated to add the go.mod to the test config, and updated the README to add an instruction on how to generate go.mod. Note that I am not copying the go.mod file into the README.

chaodaiG commented:
> For `-mod=readonly`, the container builds fine for me locally as-is, not sure what I'd change there. Why is go build in the container different in CI here?
> 
> For file-copy: It makes sense to me that we test that files we include in the README are consistent with source code, but that should not determine what is copied. For example, there are lockfiles such as `package-lock.json` for Node.js which need to be included in the build.

For file-copy, I'm fine with any change to make it smarter. One thing to keep in mind is, currently it's very tightly enforced to make sure all files needed for compilation are inspected in unit+e2e tests, if we keep separated list of what's copied and what's verified, there is a risk of these 2 lists diverged. So for any example I would recommend inline the content in README, unless it doesn't make sense at all. And in the latter case there is a way to get around it(not highly encouraged though) by running custom `cp` command, see example for Spring here: https://github.com/knative/docs/blob/533e6b7bf310bcce3ccfb628e0e736f826235844/test/sampleapp/config.yaml#L4

grayside commented:
It looks like the test is now passing, which tells me that at least some of what we were discussing was a misunderstanding that the tests require inlining all files.

chaodaiG commented:
that would be a bug if that's the case. We can have this PR pass and I'll TAL

chaodaiG commented:
/approve

grayside commented:
@chaodaiG there are files that are required for the build but are not and should not be inlined in the READMEs, for example, `package-lock.json` for Node.js. If the build leaves these out, they may pass but the result will be non-deterministic.

chaodaiG commented:
I agree that there are certain exceptions like in `package.json` case, which as I mentioned earlier can get exempted by adding a manual `cp` step in `preCommand`.

samodell commented:
/approve
/lgtm
